lib/pool.dart

```diff
@@ -121,6 +121,44 @@ class Pool {
     });
   }
 
+  /// Requests a resource for the combined duration of the [callback] and the
+  /// processing of the Stream it returns.
+  /// (E.g. a database connection is hold until the the results get processed.)
+  Stream/*<T>*/ streamWithResource/*<T>*/(Stream/*<T>*/ callback()) {
+    StreamController/*<T>*/ streamController;
+    StreamSubscription/*<T>*/ streamSubscription;
+    Completer resourceCompleter = new Completer();
+    final complete = () {
+      if (streamSubscription != null) {
+        streamSubscription.cancel();
+        streamSubscription = null;
+      }
+      if (!resourceCompleter.isCompleted) {
+        resourceCompleter.complete();
+        streamController.close();
+      }
+    };
+    streamController = new StreamController/*<T>*/(onCancel: complete);
+    withResource(() {
+      if (resourceCompleter.isCompleted) return null;
+      new Future.sync(callback).then((Stream/*<T>*/ stream) {
+        if (stream == null) {
+          complete();
+          return null;
+        }
+        streamSubscription = stream.listen(streamController.add,
+            onError: streamController.addError,
+            onDone: complete,
+            cancelOnError: true);
```

Comment:
> Added it to a new branch: https://github.com/dart-lang/pool/compare/master...isoos:stream2?expand=1

For "the stream holds a resource until it is completed", the test isn't listening to the stream, which means the done event is never getting emitted. If you add `.listen(null)` it works.

The cancel test is trickier. If a stream is canceled, it'll never actually emit a done event, but there's not currently a great way to intercept cancellations gracefully. That's probably a hole in our APIs; if you're willing to wait a day or two, I can work on a patch to the `async` package that'll add a transformer that does that nicely.

> I've also realized that I will need an easy-to-access on-completion signal to release the database connection. I've done a prototype for that here:
> https://github.com/dart-lang/pool/compare/master...isoos:stream3?expand=1

Can't you just use the returned stream's "done" event to check this? Using a callback (or a future-back) feels like a violation of the layering philosophy of this package. We generally expect the user to tell pool when resources are being used or when they're not, and this inverts that.